### ISSUE TITLE
Type narrowing tagged union consisting of classes by its tag using `match` does not work

### ISSUE THREAD

**Issue by rijenkii** [microsoft/pyright#6752]
**Describe the bug**
See also: https://github.com/python/mypy/issues/16286
Narrowing a type of a tagged union that consists of classes does not work with `match` statement, but does work with `if`s.

**Code or Screenshots**
```python
from typing import Literal

class A:
    tag: Literal["a"]
    name: str

class B:
    tag: Literal["b"]
    num: int

def f(d: A | B) -> None:
    if d.tag == "a":
        reveal_type(d)  # Type of "d" is "A"
    elif d.tag == "b":
        reveal_type(d)  # Type of "d" is "B"

def g(d: A | B) -> None:
    match d.tag:
        case "a":
            reveal_type(d)  # Type of "d" is "A | B"
        case "b":
            reveal_type(d)  # Type of "d" is "A | B"
```
(example taken from https://github.com/python/mypy/issues/16286#issuecomment-1856326305.)

**VS Code extension or command-line**
```
$ pyright --version
pyright 1.1.341
```

<details><summary>Little off-topic</summary>

Unlike mypy, pyright does correctly handle the `TypedDict` version:
```python
from typing import Literal, TypedDict

class A(TypedDict):
    tag: Literal["a"]
    name: str
    
class B(TypedDict):
    tag: Literal["b"]
    num: int

def f(d: A | B) -> None:
    if d["tag"] == "a":
        reveal_type(d)  # Type of "d" is "A"
    elif d["tag"] == "b":
        reveal_type(d)  # Type of "d" is "B"

def g(d: A | B) -> None:
    match d["tag"]:
        case "a":
            reveal_type(d)  # Type of "d" is "A"
        case "b":
            reveal_type(d)  # Type of "d" is "B"
```
(example taken from https://github.com/python/mypy/issues/16286#issue-1950150457.)

</details>

**Comment by erictraut** [comment]
This isn't a supported feature, so I don't consider this a bug. I'll switch it to an enhancement request and consider it in that light.

Note that mypy doesn't support narrowing for this case either. 

The issue here is that the subject expression is `d.tag`, but you're expecting pyright to narrow the type of `d`, which is not the same expression.

As you've noted, pyright does implement special-case logic to handle form of narrowing for conditional statements.

I'll investigate whether it's feasible to implement similar logic for pattern matching. If it doesn't negatively impact analysis performance in the general case and doesn't add too much complexity to the type checker, I think it's a worthwhile feature to consider.

**Comment by erictraut** [comment]
Out of curiosity, why don't you simply use class patterns in this case? It seems like you're jumping through unnecessary hoops to introduce tags and then key off the tag.

```python
def g(d: A | B) -> None:
    match d:
        case A():
            reveal_type(d)
        case B():
            reveal_type(d)
```

**Comment by rijenkii** [comment]
My specific use case includes pydantic and looks something like this:
```python
class A1(BaseModel):
  type: Literal["a"]
  field1: int

class A2(BaseModel):
  type: Literal["a"]
  field2: str

class B(BaseModel):
  type: Literal["b"]
  field: int

type A = A1 | A2
type AB = A | B
```

In this case checking the tag is simpler than checking the type, because I can't use `A` in `match`, I need to import both `A1` and `A2` in the main file to check, and I need to remember to update the match in case I add an `A3`.

**Comment by erictraut** [comment]
This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.342, which I just published. It will also be included in a future release of pylance.
